src/actions/profile-view.js

```diff
@@ -44,21 +55,131 @@ import type {
+ */
+export function selectLeafCallNode(
+  threadIndex: ThreadIndex,
+  sampleIndex: IndexIntoSamplesTable
+): ThunkAction<void> {
+  return (dispatch, getState) => {
+    const threadSelectors = selectorsForThread(threadIndex);
+    const filteredThread = threadSelectors.getFilteredThread(getState());
+    const callNodeInfo = threadSelectors.getCallNodeInfo(getState());
+
+    const newSelectedStack = filteredThread.samples.stack[sampleIndex];
+    const newSelectedCallNode =
+      newSelectedStack === null
+        ? -1
+        : callNodeInfo.stackIndexToCallNodeIndex[newSelectedStack];
+    dispatch(
+      changeSelectedCallNode(
+        threadIndex,
+        getCallNodePathFromIndex(
+          newSelectedCallNode,
+          callNodeInfo.callNodeTable
+        )
+      )
+    );
+  };
+}
+
+/**
+ * This function provides a different strategy for selecting call nodes. It selects
+ * a "best" ancestor call node, but also expands out its children nodes to the
+ * actual call node that was clicked. See findBestAncestorCallNode for more
+ * on the "best" call node.
+ */
+export function selectBestAncestorCallNodeAndExpandCallTree(
+  threadIndex: ThreadIndex,
+  sampleIndex: IndexIntoSamplesTable
+): ThunkAction<boolean> {
+  return (dispatch, getState) => {
+    const threadSelectors = selectorsForThread(threadIndex);
+    const fullThread = threadSelectors.getRangeFilteredThread(getState());
+    const filteredThread = threadSelectors.getFilteredThread(getState());
+    const unfilteredStack = fullThread.samples.stack[sampleIndex];
+    const callNodeInfo = threadSelectors.getCallNodeInfo(getState());
+
+    if (unfilteredStack === null) {
+      return false;
+    }
+
+    const { callNodeTable, stackIndexToCallNodeIndex } = callNodeInfo;
+    const sampleCallNodes = getSampleCallNodes(
+      filteredThread.samples,
+      stackIndexToCallNodeIndex
+    );
+    const clickedCallNode = sampleCallNodes[sampleIndex];
+    const clickedCategory = fullThread.stackTable.category[unfilteredStack];
+
+    if (clickedCallNode === null) {
+      return false;
+    }
+
+    const sampleCategories = getSampleCategories(
+      fullThread.samples,
+      fullThread.stackTable
+    );
+    const bestAncestorCallNode = findBestAncestorCallNode(
+      callNodeInfo,
+      sampleCallNodes,
+      sampleCategories,
+      clickedCallNode,
+      clickedCategory
+    );
+
+    // In one dispatch, change the selected call node to the best ancestor call node, but
+    // also expand out to the clicked call node.
+    dispatch(
+      changeSelectedCallNode(
+        threadIndex,
+        // Select the best ancestor call node.
+        getCallNodePathFromIndex(bestAncestorCallNode, callNodeTable),
+        // Also expand the children nodes out further below it to what was actually
+        // clicked.
+        getCallNodePathFromIndex(clickedCallNode, callNodeTable)
+      )
+    );
+    return true;
```

Comment:
Oh, the return value of this thunk function is synchronously propagated to the caller of the bound action? I always thought thunks came with some asynchronicity. Well, I learned something new today.

Reply:
Yeah, I had to look it up to double check. The type definitions helped me realize what was going on a little bit more.